src/VisualStudio/Core/Def/Implementation/TableDataSource/Suppression/VisualStudioDiagnosticListSuppressionStateService.cs

```diff
@@ -0,0 +1,253 @@
+            {
+                isRoslynEntry = false;
+                isSuppressedEntry = false;
+                isCompilerDiagnosticEntry = false;
+                return false;
+            }
+
+            isRoslynEntry = true;
+            isSuppressedEntry = diagnosticData.IsSuppressed;
+            isCompilerDiagnosticEntry = SuppressionHelpers.IsCompilerDiagnostic(diagnosticData);
+            return true;
+        }
+
+        private static AbstractTableEntriesSnapshot<DiagnosticData> GetEntriesSnapshot(ITableEntryHandle entryHandle)
+        {
+            int index;
+            return GetEntriesSnapshot(entryHandle, out index);
+        }
+
+        private static AbstractTableEntriesSnapshot<DiagnosticData> GetEntriesSnapshot(ITableEntryHandle entryHandle, out int index)
+        {
+            ITableEntriesSnapshot snapshot;
+            if (!entryHandle.TryGetSnapshot(out snapshot, out index))
+            {
+                return null;
+            }
+
+            return snapshot as AbstractTableEntriesSnapshot<DiagnosticData>;
+        }
+
+        public ImmutableArray<DiagnosticData> GetItems(bool selectedEntriesOnly, bool isAddSuppression, bool isSuppressionInSource, CancellationToken cancellationToken)
+        {
+            var builder = ImmutableArray.CreateBuilder<DiagnosticData>();
+            var entries = selectedEntriesOnly ? _tableControl.SelectedEntries : _tableControl.Entries;
+            foreach (var entryHandle in entries)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                DiagnosticData diagnosticData = null;
+                int index;
+                var roslynSnapshot = GetEntriesSnapshot(entryHandle, out index);
+                if (roslynSnapshot != null)
+                {
+                    diagnosticData = roslynSnapshot.GetItem(index)?.Primary;
+                    if (diagnosticData != null && diagnosticData.HasTextSpan)
+                    {
+                        if (isAddSuppression)
+                        {
+                            // Compiler diagnostics can only be suppressed in source.
+                            if (!diagnosticData.IsSuppressed &&
+                                (isSuppressionInSource || !SuppressionHelpers.IsCompilerDiagnostic(diagnosticData)))
+                            {
+                                builder.Add(diagnosticData);
+                            }
+                        }
+                        else if (diagnosticData.IsSuppressed)
+                        {
+                            builder.Add(diagnosticData);
+                        }
+                    }
+                }
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private static void UpdateSelectedItems(bool added, ref int count)
+        {
+            if (added)
+            {
+                count++;
```

Comment:
is there a point where it make sure the bookkeeping is right or assert that make sure the bookkeeping is right?